### PR TITLE
Add attrs namespace

### LINKED ISSUE #487: [RFC] Inconvenient defaults?
As part of “Project`import attrs`” (see also #408) we get a unique opportunity to change default options/behaviors that grew over the years but couldn't be fixed due to backward-compatibility restrictions.

The rough plan is create a new, friendlier API on top of `@attr.s` and `attr.ib()` that won't go anywhere.

The following is cast in stone:

- `auto_attribs=True`
- consistent (Python 3-style) hashing behavior on every Python version
- #428 

The following would be really nice to have:

- #368 but it might need too much thinking/research

What else is bugging you?

***

One thing that I just can't make up my mind is related to #223: should we make `slots=True` the default? I’m quite confident that in 99,9% of cases it's the right thing to do and will guide people to write better classes.

However on the other hand, our approach of rewriting classes breaks in certain scenarios, usually involving metaclasses.

So the question is, whether we want to tolerate a higher rate of bogus bug reports/help requests or make the default case nicer?

I welcome your input.

***

Finally a controversial idea: we could make `import attrs` Python 3 only. There isn't much baggage we'd get rid of but there is _some_ and 2020 is less than a year ahead. It would also allow us to embrace enums as part of our API.

### LINKED ISSUE #705: Better defaults for asdict/astuple
- Would it be acceptable to change the default value for `retain_collection_types` to `True`? This would reduce the likelihood of this issue to appear unexpectedly. It is not a full fix though.

- Would it be ok to use `tuple` as default collection type, when `retain_collection_types=False`? I would expect this to completely resolve the issue.

_Originally posted by @tovrstra in https://github.com/python-attrs/attrs/issues/646#issuecomment-682167908_

### PR DESCRIPTION
**Pull request by hynek**
# Summary

ROCK’N’ROLL

Fixes #487, fixes #705,

I don't think it's practical to already update all the docs to the new namespace. Especially since it's provisional for now.

# Pull Request Check List

- [x] Added **tests** for changed code.
- [x] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/main/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x] ...and used in the stub test file `tests/typing_example.py`.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/main/src/attr/__init__.py) file.
- [x] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/main/changelog.d).

### CODE REVIEW

File: src/attrs/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+)
+
+# Because we need to type our own stuff, we have to make everything from
+# attr explicitly public too.
+from attr import __author__ as __author__
```
Comment (hynek) [inline src/attrs/__init__.pyi:14]: @euresti is there a better way to deal with this stuff maybe?

Comment (wbolster) [reply src/attrs/__init__.pyi:14]: would `__all__` help perhaps? some tools (flake8 afaik) look at it and will not warn about ‘unused imports’ for anything appearing there. not sure if similar things are in place for tools (mypy etc) looking at .pyi files.

Comment (hynek) [reply src/attrs/__init__.pyi:14]: help with what exactly? i'd like to elide that whole block :)

the problem is that i need to have type hints for asdict/astuple and with that I can't use tacitly those from attr/__init__.pyi and have to import them here

File: src/attr/__init__.pyi [hunk 2]
```diff
@@ -447,6 +447,7 @@ def make_class(
 # these:
 # https://github.com/python/mypy/issues/4236
 # https://github.com/python/typing/issues/253
+# XXX: remember to fix attrs.asdict/astuple too!
```
Comment (wbolster) [inline src/attr/__init__.pyi:450]: is this for now or for later?

Comment (hynek) [reply src/attr/__init__.pyi:450]: later – its referring to the changes to mypy/typing that once they're done, we need to fix both places :)

File: src/attrs/__init__.pyi (hunk 1 above)
Comment (wbolster) [reply src/attrs/__init__.pyi:14]: ack, i'm missing some context here. i was referring to the `x as x` repetition

File: src/attr/__init__.pyi (hunk 2 above)
Comment (wbolster) [reply src/attr/__init__.pyi:450]: ack, i wondered since both links point to _closed_ issues

**Comment by Tinche** [comment]
If `attrs` is the future, shouldn't we move all the code over there and import it in `attr`?

**Comment by hynek** [comment]
> If `attrs` is the future, shouldn't we move all the code over there and import it in `attr`?

Technically yes, but that would be a breaking change whose blast radius I can't fathom.

This way, even MyPy keeps working out of the box.

File: src/attr/__init__.pyi (hunk 2 above)
Comment (hynek) [reply src/attr/__init__.pyi:450]: I hope sOmEoNe tackles it then 😇

File: src/attrs/__init__.pyi (hunk 1 above)
Comment (hynek) [reply src/attrs/__init__.pyi:14]: according to `attr/__init__.pyi` it’s necessary to make them public and in my experiments it still seems to hold true 🤔

**Comment by Tinche** [comment]
```
>>> from attrs.converters import optional
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'attrs.converters'
```

**Comment by hynek** [comment]
ugh crap, I guess we have to mirror all packages into the new namespace?

**Comment by hynek** [comment]
fe3c53e is ugly but should do the job?

**Comment by wbolster** [comment]
> [fe3c53e](https://github.com/python-attrs/attrs/commit/fe3c53eda43346f205145193f5ad4b871f60abfb) is ugly but should do the job?

fwiw, mypy is generally (and rightfully) not happy about `*` imports

**Comment by hynek** [comment]
I played with `reveal_type` and it _seems_ to work? Did they improve or am I missing something?

**Comment by wbolster** [comment]
ah ok, i think the separate `.pyi` file makes it work in this case.

in general, i don't think it works well with (modern) embedded annotations directly in the source code.

**Comment by hynek** [comment]
I'm open to suggestions for improvement, but one more repetition would mean one more place we could forget about. :-/

**Comment by euresti** [comment]
Do you need the pyi file or would mypy pick up the types from the py file?

**Comment by hynek** [comment]
We need overloads, so I'm reasonably sure we need `pyi` files?

**Comment by wbolster** [comment]
imho, as a mostly bystander and casual contributor (aka ignore me), doing these things at the same time would be perfect timing: 🙃

- introducing new package name `attrs`
- becoming py37+ only (only supported pythons from next week)

…but i suspect you may hold different opinions on that, based on [py2 notes in the docs](https://www.attrs.org/en/stable/python-2.html)

**Comment by hynek** [comment]
While that would solve other pain points, overloads are AFAIK something that you need stub files for still. :)

**Comment by wbolster** [comment]
> We need overloads, so I'm reasonably sure we need `pyi` files?

overloads work fine in source code, see e.g.
https://github.com/wbolster/jsonlines/blob/master/jsonlines/jsonlines.py

**Comment by euresti** [comment]
Yeah. We could definitely move the stubs into the py files.  Though attrs has quite a bit of "fake" types that we wouldn't want mypy checking.  It could be an interesting project.

**Comment by euresti** [comment]
Is the expectation that `import attr` is frozen and will not be receiving any new features that `import attrs` might be?  Or is anything available in `attrs` gonna be available in `attr`?

**Comment by wbolster** [comment]
> Though attrs has quite a bit of "fake" types that we wouldn't want mypy checking.

`@typing.type_check_only` may help here: [docs](https://docs.python.org/3/library/typing.html#typing.type_check_only)

**Comment by hynek** [comment]
This all sounds very promising, but this diff is already way too big. Can I somehow convince y'all to help me with inline types for the next release?

attrs 2.7 downloads seem to be down to 2.22% so I might change my mind soon-ish:

| category | percent |  downloads |
|:---------|--------:|-----------:|
| 3.7      |  40.10% | 19,465,561 |
| 3.8      |  21.25% | 10,315,653 |
| 3.6      |  12.75% |  6,189,331 |
| null     |  10.43% |  5,063,048 |
| 3.9      |   9.15% |  4,442,703 |
| 2.7      |   2.22% |  1,079,185 |
| 3.10     |   2.06% |    998,378 |
| 3.5      |   2.01% |    974,174 |
| 3.4      |   0.03% |     13,247 |
| 3.11     |   0.01% |      3,357 |
| 2.6      |   0.00% |          8 |
| 3.3      |   0.00% |          6 |
| 3.2      |   0.00% |          4 |
| Total    |         | 48,544,655 |

Date range: 2021-12-01 - 2021-12-20

**Comment by Tinche** [comment]
I'm going to follow attrs lead in cattrs (big surprise), and have a PR ready. What do I say in the changelog though? Are we recommending new code to import from attrs/cattrs instead? Do we ever plan on removing attr/cattr?

**Comment by hynek** [comment]
Well, you can see what I've said in https://github.com/python-attrs/attrs/pull/887/files#diff-826ab996c410e75f5930990d312d073db10300312fc9fabe2f01fb327a007bde

I'm leaving it provisional for one release, but I think changing the docs to primarily use the new APIs is a strong indicator. ;)

Since we had almost 90 million downloads per month with `attr` in Place, I don't remove it, ever. I think I'll eventually switch the import direction but that's it. 🤔

**Comment by hynek** [comment]
I don't think I can make anyone review a +830/-368 PR so here we go and let's hope for the best. Merry Xmas to me. 🤪

File: docs/api.rst [hunk 3]
```diff
@@ -341,99 +440,99 @@ Validators can be globally disabled if you want to run them only in development
 Validators
 ----------
 
-``attrs`` comes with some common validators in the ``attrs.validators`` module:
+``attrs`` comes with some common validators in the ``attrs.validators`` module.
+All objects from ``attrs.converters`` are also available from ``attr.converters``.
```
Comment (sanjioh) [inline docs/api.rst:444]: @hynek Hi! I just wanted to let you know that this line about converters does not probably belong here (it looks like a duplicate of a line down below). Thanks for attrs and for its wonderful docs!

Comment (hynek) [reply docs/api.rst:444]: thanks, fixed in e188af7bc4df99eb84772b0ad0f80c4e11f740b4!

**Pull request merged**